File: (03-19-2024)Best Time to Buy and Sell Stock/joshlin/solution.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # time: O(n), space: O(1)
        if len(prices) < 2:
            return 0
        max_profit, min_stock = float('-inf'), prices[0]
        for price in prices:
            max_profit = max(max_profit, price - min_stock)
            min_stock = min(min_stock, price)
        return max_profit

File: (03-19-2024)Best Time to Buy and Sell Stock/joshlin/solution.ts
function maxProfit(prices: number[]): number {
    // time: O(n), space: O(1)
    if (prices.length < 2) return 0;
    let max_profit = 0;
    let min_stock = prices[0];
    for (const price of prices) {
        const profit = price - min_stock;
        max_profit = profit > max_profit ? profit : max_profit;
        min_stock = price < min_stock ? price : min_stock;
    };
    return max_profit;
};